### ISSUE TITLE
the 'push' TravisCI job is not useful for pull requests

### ISSUE THREAD

**Issue by prometheanfire** [rcbops/rpc-openstack#972]
##### What version (commit) do you encounter this issue on?
all

##### What type of issue is this (bug, feature, documentation, regression)?
bug

##### What is the expected behavior?
Only tests of the resulting final code (merge/rebase) be done.

##### What is actually happening?
tests being run on the branch with no merge

##### How can we reproduce this issue?
make a pr from 15fac66f4871953bded33f899278b1090dee6439 and see what happens

##### How can we verify your issue is addressed?
only testing of the final resultant code is run

##### If this is a large issue, what can be done immediately to alleviate the issue while working on a longer-term solution?
disable the push travisci job

##### What is your evaluation on the importance of this issue, and why?
this is obviously a stop the presses level of bug

###### Remove branches (lines) you do not wish to target.
- [ ] Master
- [ ] liberty-12.1
- [ ] liberty-12.0
- [ ] kilo

**Comment by sigmavirus24** [comment]
We're not using circleci

**Comment by prometheanfire** [comment]
s/circle/travis

**Comment by sigmavirus24** [comment]
If you look at the [settings for Travis CI](https://travis-ci.org/rcbops/rpc-openstack/settings) you'll see we either build all pushes or no pushes at all. As a team, we've discussed this several times and decided we don't want to ignore pushes to a new branch (e.g., liberty-12.1) that do not go through a pull request simply to avoid this problem. Are there new problems/solutions to this or are we rehashing the same discussions again with nothing but refreshed frustrations?

**Comment by prometheanfire** [comment]
So, we can't remove the check entirely?

**Comment by sigmavirus24** [comment]
It is possible to remove the check. Whether we should remove the check is a different discussion (one we have had at least once prior). The conversation boiled down to, roughly,:

- Any push we do directly to a branch (e.g., a stable branch) will not be tested on push which means all PRs will start failing when trying to merge to that branch (without explanation)

- Travis does not re-run tests when master updates, so a merged build might "pass" and then after merging might fail. We would know that until GitHub "pushes" the code to the branch that the PR merges to resulting in a branch that new/updated PRs fail against unexpectedly.

Both of those situations are easy to run into, and I think not having an early notice of a failed merge is more harmful than the frustrating situation we can run into that you've documented.

That's my 2¢ though. I'll let others chime in.

**Comment by prometheanfire** [comment]
* I don't think disabling will cause all PRs to fail testing, that's a separate test.

* Travis already does not re-run tests when the target branch udpates, so old PRs can still show as good.  This is something we are already dealing with.

The push check does not give us an early notice of a failed merge, all it does is give us a false positive when there's a error in a feature branch, not in the final merged product.

**Comment by sigmavirus24** [comment]
> I don't think disabling will cause all PRs to fail testing, that's a separate test.

That's not what I said.

> Travis already does not re-run tests when the target branch udpates, so old PRs can still show as good. This is something we are already dealing with.

We do not seem to be triggering new merged builds after each PR is merged to a branch. How exactly are we already dealing with this problem?

> The push check does not give us an early notice of a failed merge

Not of a failed merge. It gives us feedback of a successful merge that will cause new merges to fail.

**Comment by prometheanfire** [comment]
> * Any push we do directly to a branch (e.g., a stable branch) will not be tested on push which means all PRs will start failing when trying to merge to that branch (without explanation)

This is handled by the merge job already, we don't push directly to those branches.

> > Travis already does not re-run tests when the target branch udpates, so old PRs can still show as good. This is something we are already dealing with.

> We do not seem to be triggering new merged builds after each PR is merged to a branch. How exactly are we already dealing with this problem?

We are not dealing with it, that's a separate issue, this issue is about removing a false positive, please stay on topic.

> > The push check does not give us an early notice of a failed merge

> Not of a failed merge. It gives us feedback of a successful merge that will cause new merges to fail.

How does this feedback cause new merges to fail?  Is it adding code to the repo?

**Comment by sigmavirus24** [comment]
Okay, let's establish some *facts* and work from there.

Facts:

- GitHub pushes to a branch on merge with the commit from the PR and the merge commit.
- Pushes to branches by arbitrary people are allowed for everything except `kilo` and `master`.
- We create branches on this repository for pull requests.
- Not all branches become pull requests, some are used for managing separate versions of a release (e.g., `liberty`, `liberty-12.1`)
- New branches are not created by a merge job or GitHub
- Pull Requests have 5 different checks run against them, two of which are "Travis-CI Push" and "Travis-CI PR"
- Branches that have changes pushed to them (either by people or by GitHub) have "Travis-CI Push" ran against them
- Branches are **not** required to be up-to-date with master because we found that to be an onerous requirement
- Travis-CI run the "Push" and "PR" jobs once
- As long as a PR can be merged (according to GitHub), and all checks pass, we will merge it
- Changes merged to a branch in the past have been "mergeable" but not necessarily correct once a new PR is merged to that branch in the past (more than once)

So now let's use our aggregated facts above:

- If we have branch `A` that has recently had PR 1 merged into it and we merge PR 2 into it without issuing *new* merged checks, `A` may end up in a state where all new PRs against `A` fail because `A` itself is failing.
- If we disable "Travis-CI Push" for the project, no check will run when PR 2 is merged. (This means that we won't know `A` is broken until someone submits PR 3.)
- If we disable "Travis-CI Push" and branch off of an already broken branch, now we will have two broken branches because the person who pushed the new branch will be unaware that their base branch was broken (until a new PR is sent to either or both branches).

**In your opinion, would it be better to know that a branch is broken sooner or later?**

Further, do you dispute the above facts and conclusions?

---

> We are not dealing with it, that's a separate issue, this issue is about removing a false positive, please stay on topic.

This discussion is on topic, because we're not just "removing a false positive" we're also removing an early warning system.

**Comment by prometheanfire** [comment]
> * Changes merged to a branch in the past have been "mergeable" but not necessarily correct once a new PR is merged to that branch in the past (more than once)

do you have examples?

> * If we disable "Travis-CI Push" for the project, no check will run when PR 2 is merged. (This means that we won't know A is broken until someone submits PR 3.)

This is already the case.  You seem to be under the impression that once a pr is merged a test (that we view) is run again, that would be useful if true, but it's not (from what I can see).

> * If we disable "Travis-CI Push" and branch off of an already broken branch, now we will have two broken branches because the person who pushed the new branch will be unaware that their base branch was broken (until a new PR is sent to either or both branches).

This is true, branching off of something already broken will have that branch stay broken.  Since we typically (almost always in my case at least) branch off of 'stable' branches, which have restricted push access, they should be considered 'safe' because only PRs that pass the PR test will be merged.

As far as branching off of a branch that's already broken... Can you tell me, without submitting a PR, if a branch is passing?  This seems to be something you think can easily be done.

**Comment by sigmavirus24** [comment]
> This is already the case. You seem to be under the impression that once a pr is merged a test (that we view) is run again, that would be useful if true, but it's not (from what I can see).

You're ignoring the fact that Travis runs tests for *every* push. You can see that [here](https://travis-ci.org/rcbops/rpc-openstack/branches) and if you want, I can configure it so that every push failure is sent to your email address. We can have it sent wherever we want (given it's currently unconfigured and has disappeared since I last saw it). (You can see that "stable" `liberty-12.1` has two failed builds at that link too.)

> As far as branching off of a branch that's already broken... Can you tell me, without submitting a PR, if a branch is passing? 

Refer to my previous paragraph.

**Comment by prometheanfire** [comment]
If travis runs test for every push, that's great, then we should probably set up that email to mail somewhere useful.  I'm not sure anyone but you looks at that when branching, perhaps having a status/badge might be more helpful?

**Comment by hughsaunders** [comment]
As an alternative, we could run all the jobs via jenkins (functional and lints), have jenkins initially test each pr and report the job status on the commit as it does currently, then have a keyword (such as :shipit: ) that jenkins looks for which kicks off an infra-style verify job that runs the tests again and merges if they pass?  Then we have always run a test immediately pre-merge. Its not entirely free of race conditions but may be better than the current situation?

**Comment by sigmavirus24** [comment]
@hughsaunders that may be better if we can steal some of the work that openstack-infra does (e.g., queuing changes to merge instead of doing them in parallel).

**Comment by prometheanfire** [comment]
That'd likely help at least somewhat

**Comment by prometheanfire** [comment]
We are going to be using a status page which will track the status of branches.  So we keep the job. ezpz
